File: webapp/Locust/Scripts/app/services/TagService.js
(function (angular, undefined) {
    'use strict';

    // Get services module
    var app = angular.module(appName);

    app.service('TagService', ['$resource', function ($resource) {

        var noop = angular.noop;
        var resource = $resource('/api/tag/:action/:id',
            { id: '@Id' }, {
                Save: {
                    method: 'POST'
                },
                Update: {
                    method: 'PUT'
                },
                Delete: {
                    method: 'DELETE',
                    params: {
                        id: "@id"
                    }
                },
                Get: {
                    method: 'GET',
                    isArray: false
                },
                GetAll: {
                    method: 'GET',
                    isArray: true
                },

                GetTestCaseTags: {
                    method: 'GET',
                    isArray: true
                },
                GetTestScenarioTags: {
                    method: 'GET',
                    isArray: true
                },
                GetTestProcedureTags: {
                    method: 'GET',
                    isArray: true
                },
                GetTestSuplementalTags: {
                    method: 'GET',
                    isArray: true
                }
            }),

            self = this;

        self.GetAll = function (success, error) {
            return resource.GetAll({ action: 'GetAll' }, success || noop, error || noop);
        };

        self.GetTestCaseTags = function (idtc, success, error) {
            return resource.GetTestCaseTags({ action: 'GetTestCaseTags', idtc: idtc }, success || noop, error || noop);
        };

        self.GetTestScenarioTags = function (idts, success, error) {
            return resource.GetTestCaseTags({ action: 'GetTestScenarioTags', idts: idts }, success || noop, error || noop);
        };

        self.GetTestProcedureTags = function (idtp, success, error) {
            return resource.GetTestCaseTags({ action: 'GetTestProcedureTags', idtp: idtp }, success || noop, error || noop);
        };

        self.GetTestSuplementalTags = function (idstp, success, error) {
            return resource.GetTestCaseTags({ action: 'GetTestSuplementalTags', idstp: idstp }, success || noop, error || noop);
        };

        self.Get = function (id, success, error) {
            return resource.Get({ action: 'Get', id: id }, success || noop, error || noop);
        };

        self.Delete = function (idTag, idtc, idts, idtp,idstp, success, error) {
            return resource.Delete({ action: 'Delete', idTag: idTag, idtc: idtc, idts: idts, idtp: idtp, idstp: idstp }, success || noop, error || noop);
        };

        self.Save = function (entity, idtc, idts, idtp, idstp, success, error) {
            return resource.Save({ action: 'Save', idtc: idtc, idts: idts, idtp: idtp, idstp: idstp }, entity, success || noop, error || noop);
        };

        self.Update = function (entity, success, error) {
            return resource.Update({ action: 'Update' }, entity, success || noop, error || noop);
        };

    }]);
})(angular);

File: webapp/Locus.Core/Services/TagService.cs
using Locus.Core.Models;
using Locus.Core.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Locus.Core.Services
{
    //Interface which contains methods for each CRUD operation
    public interface ITagService
    {

        Tag Save(Tag tag, int id, int idts, int idtp, int idstp);

        Tag Update(Tag tag);

        Tag Delete(int idtag, int idtc, int idts, int idtp, int idstp);

        List<Tag> GetAll();

        Tag Get(int idtag);

        List<Tag> GetTestCaseTags(int idtc);

        List<Tag> GetTestScenarioTags(int idts);

        List<Tag> GetTestProcedureTags(int idtp);

        List<Tag> GetTestSuplementalTags(int idstp);
    }

    public class TagService : ITagService
    {
        //Instance of IProjectRepository
        private readonly ITagRepository _tagRepository;

        public TagService(ITagRepository tagRepository)
        {
            _tagRepository = tagRepository;
        }

        public Tag Delete(int idtag, int idtc, int idts, int idtp, int idstp )
        {
            return _tagRepository.Delete(idtag, idtc, idts, idtp, idstp);
        }

        public Tag Get(int idtag)
        {
            return _tagRepository.Get(idtag);
        }

        public List<Tag> GetAll()
        {
            return _tagRepository.GetAll();
        }

        public List<Tag> GetTestCaseTags(int idtc)
        {
            return _tagRepository.GetTestCaseTags(idtc);
        }

        public List<Tag> GetTestProcedureTags(int idtp)
        {
            return _tagRepository.GetTestProcedureTags(idtp);
        }

        public List<Tag> GetTestScenarioTags(int idts)
        {
            return _tagRepository.GetTestScenarioTags(idts);
        }

        public List<Tag> GetTestSuplementalTags(int idstp)
        {
            return _tagRepository.GetTestSuplementalTags(idstp);
        }

        public Tag Save(Tag tag, int idtc, int idts, int idtp, int idstp)
        {
            return _tagRepository.Save(tag, idtc, idts, idtp, idstp);
        }

        public Tag Update(Tag tag)
        {
            return _tagRepository.Update(tag);
        }
    }
}
